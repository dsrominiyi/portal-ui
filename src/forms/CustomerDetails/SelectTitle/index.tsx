import React, { FunctionComponent, Fragment } from 'react';

import QuestionBox from '../../../components/QuestionBox';
import Dropdown from '../../../components/Dropdown';
import MultiChoice from '../../../components/MultiChoice';
import SubQuestion from '../../../components/SubQuestion';

import { otherTitleOptions } from '../../../helpers/optionLists';

export interface Props {
  title?: string;
  hasError?: boolean;
  onChange(value: string): void;
}

const titleList = ['Mr', 'Mrs', 'Miss', 'Ms', 'Other'];
const standardTitles = [...titleList];
const titleOther = standardTitles.pop(); // intentionally remove 'Other' from standardTitles

export const SelectTitle: FunctionComponent<Props> = ({
  title = '',
  hasError,
  onChange
}: Props) => (
  <Fragment>
    <QuestionBox
      text="Title"
      hasError={hasError}
      errorText="Please confirm title"
      inputFields={[
        <MultiChoice
          key="0"
          options={titleList}
          selectedOption={!title || standardTitles.includes(title) ? title : titleOther}
          onClick={option => onChange(option)}
          attributes={{ 'test-id': 'select-standard-title' }}
        />
      ]}
    />
    <SubQuestion
      visible={!!title && !standardTitles.includes(title)}
      items={[
        <QuestionBox
          key="0"
          hasError={hasError}
          inputFields={[
            <Dropdown
              key="0"
              label="Please select title"
              options={otherTitleOptions}
              selected={otherTitleOptions.includes(title) ? title : undefined}
              onChange={value => onChange(String(value))}
              attributes={{ 'test-id': 'select-alternative-title' }}
            />
          ]}
        />
      ]}
    />
  </Fragment>
);

SelectTitle.displayName = 'SelectTitle';
export default SelectTitle;
