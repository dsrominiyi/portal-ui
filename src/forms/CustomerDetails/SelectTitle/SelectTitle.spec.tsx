import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import SelectTitle, { Props } from '.';

let fields: ShallowWrapper<Props>;
let standardTitles: any;
let otherTitles: any;
const onChange = jest.fn();

const update = () => {
  [standardTitles, otherTitles] = (fields.props() as any).children;
};

describe('SelectTitle', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fields = shallow(<SelectTitle onChange={onChange} />);
    update();
  });

  test('SelectTitle matches snapshot', () => {
    expect(fields).toMatchSnapshot();
  });

  test('selecting a standard title calls onChange with the value', () => {
    const title = 'Mr';
    standardTitles.props.inputFields[0].props.onClick(title);

    expect(onChange).toHaveBeenCalledWith(title);
  });

  test('the other titles dropdown is made visible if title is "Other"', () => {
    expect(otherTitles.props.visible).toBe(false);

    fields.setProps({ title: 'Other' });
    update();

    expect(otherTitles.props.visible).toBe(true);
  });

  test('selecting an alternative title calls onChange with the value', () => {
    const title = 'Mx';
    otherTitles.props.items[0].props.inputFields[0].props.onChange(title);

    expect(onChange).toHaveBeenCalledWith(title);
  });

  test('"Other" under the standard titles is selected if an alternative title from the dropdown is selected', () => {
    fields.setProps({ title: 'Mx' });
    update();

    expect(standardTitles.props.inputFields[0].props.selectedOption).toBe('Other');
  });
});
