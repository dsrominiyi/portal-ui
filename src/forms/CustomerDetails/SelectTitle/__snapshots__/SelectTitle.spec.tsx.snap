// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectTitle SelectTitle matches snapshot 1`] = `
<Fragment>
  <QuestionBox
    errorText="Please confirm title"
    inputFields={
      Array [
        <MultiChoice
          attributes={
            Object {
              "test-id": "select-standard-title",
            }
          }
          onClick={[Function]}
          options={
            Array [
              "Mr",
              "Mrs",
              "Miss",
              "Ms",
              "Other",
            ]
          }
          selectedOption=""
        />,
      ]
    }
    text="Title"
  />
  <SubQuestion
    items={
      Array [
        <QuestionBox
          inputFields={
            Array [
              <Dropdown
                attributes={
                  Object {
                    "test-id": "select-alternative-title",
                  }
                }
                label="Please select title"
                onChange={[Function]}
                options={
                  Array [
                    "Mx",
                    "Master",
                    "Brig.",
                    "Bro.",
                    "Capt.",
                    "Chf.",
                    "Col.",
                    "Comdr.",
                    "Comm.",
                    "Cllr.",
                    "Dame",
                    "Dr.",
                    "Exct.",
                    "F/Lt.",
                    "Gen.",
                    "Hon.",
                    "Judge",
                    "Lady",
                    "Lt.",
                    "Lord",
                    "Maj.",
                    "Prof",
                    "Rabbi",
                    "Rev.",
                    "Rt Rev",
                    "Sir",
                    "Sister",
                    "Sq Ldr",
                    "Rt Hon",
                    "V. Rev",
                    "W/Cdr",
                  ]
                }
              />,
            ]
          }
        />,
      ]
    }
    visible={false}
  />
</Fragment>
`;
